### PR TITLE
Feat/enable hooks sor

### LINKED ISSUE #809: SOR Add hooks support
- Any Balancer supported hook should be added to Balancer Maths package which will handle calculations given relevant pool and hook state.
- Any Balancer supported hook should also be added to API (by API team) which will handle collecting/storing relevant hook state.
- Pools with non-supported hooks should be ignored by SOR.
- In the current SOR implementation the BalancerMaths library is also being used with pool state provided from API. Hook state should be passed in a similar manner.
- Ideally we want to structure this in a way that makes this easy to do as we will likely have a number of hooks to support (doc of process would be helpful)

I'd suggest starting by adding support for the ExitFee hook. This has already been added to maths package and I believe Jarek has already added support in API.
See Balancer Maths [ExitFee hook test](https://github.com/balancer/balancer-maths/blob/main/typescript/test/hooks/exitFee.test.ts) for example of hook state required.
Currently maths uses `hookType`(see [here](https://github.com/balancer/balancer-maths/blob/main/typescript/test/hooks/exitFee.test.ts#L7)) field on pool state to map a pool<>hook. Currently a string name. This may not match API implementation and we can figure out best approach.

### PR DESCRIPTION
**Pull request by mkflow27**
This pr will enable hooks in the SOR. The key changes/additions are:

- Pools are constructed with Liquidity management to check if the sor can build a route via single token operations
- HookState depending on the hook associated with the given pool

### CODE REVIEW

File: modules/sor/sorV2/lib/utils/helpers.ts [hunk 1]
```diff
@@ -71,3 +74,53 @@ export function getOutputAmount(paths: PathWithAmount[]): TokenAmount {
...
+
+    
+    if (pool.hook.name === 'ExitFee') {
+        // api for this hook is an Object with removeLiquidityFeePercentage key & fee as string
+        const dynamicData = pool.hook.dynamicData as { removeLiquidityFeePercentage: string };
+
+        return {
+            tokens: pool.tokens.map((token: { address: string }) => token.address),
+            // ExitFeeHook will always have dynamicData as part of the API response
+            removeLiquidityHookFeePercentage: percentageStringToBigInt(dynamicData.removeLiquidityFeePercentage),
```
Comment (brunoguerios) [inline modules/sor/sorV2/lib/utils/helpers.ts]: This is actually error prone because you're calculating with Number and then casting to BigInt. 
Better to use `parseEther` on this case where it's always 18 decimals.
For reference, `parseUnits` does the same thing, but with a custom number of decimals

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts [hunk 2]
```diff
@@ -156,6 +179,13 @@ export class StablePool implements BasePoolV3 {
         let calculatedAmount: bigint;
 
         if (tIn.token.isSameAddress(this.id)) {
+            // if liquidityManagement.disableUnbalancedLiquidity is true return 0
+            // as the pool does not allow unbalanced operations. 0 return marks the
+            // route as truly unfeasible route.
+            if (this.liquidityManagement.disableUnbalancedLiquidity) {
+                return TokenAmount.fromRawAmount(tOut.token, 0n);
+            }
```
Comment (brunoguerios) [inline modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts:185]: This seems to me more like a pool math/config that should be handled by balancer-maths. 
Any specific reason you kept it here?

File: test/factories/prismaHook.factory.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+export const hookFactory = PrismaHookFactory.define(({ params }) => {
+    const hookAddress = params?.address ?? createRandomAddress();
+
+    return {
+        address: hookAddress,
+        name: params.name || 'Test Hook',
+        enableHookAdjustedAmounts: params?.enableHookAdjustedAmounts ?? false,
+        shouldCallAfterSwap: params?.shouldCallAfterSwap ?? false,
+        shouldCallBeforeSwap: params?.shouldCallBeforeSwap ?? false,
```
Comment (brunoguerios) [inline test/factories/prismaHook.factory.ts]: Factory works in a way that you can simply state the default values for each parameter.
They will be used only if the parameter was not overwritten on `params`.
In other words, you can do this:
```suggestion
    return {
        address: createRandomAddress(),
        name: 'Test Hook',
        enableHookAdjustedAmounts: false,
        shouldCallAfterSwap: false,
        shouldCallBeforeSwap: false,
        ...
```
and you'll have the same result.

File: modules/sor/sorV2/lib/utils/helpers.ts [hunk 4]
```diff
@@ -71,3 +74,53 @@ export function getOutputAmount(paths: PathWithAmount[]): TokenAmount {
     const amounts = paths.map((path) => path.outputAmount);
     return amounts.reduce((a, b) => a.add(b));
 }
+
+export function returnHookDataAccordingToHookName(pool: any): HookState | undefined {
```
Comment (brunoguerios) [inline modules/sor/sorV2/lib/utils/helpers.ts]: Not sure if you're looking for suggestions on function names, but I'd go with `getHookState` to follow the existing `getPoolState` pattern.

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts (hunk 2 above)
Comment (brunoguerios) [reply modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts:185]: Just sharing that I talked to @johngrantuk and he feels this is better suited here in the backend code instead of balancer-maths. Let's leave as is 😉

Comment (mkflow27) [reply modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts:185]: @brunoguerios I canno't find the original chat message but I recall you talking to john and agreeing that this can be left in here instead of moving it to balancer-maths.

File: modules/sor/sorV2/lib/utils/helpers.ts (hunk 4 above)
Comment (mkflow27) [reply modules/sor/sorV2/lib/utils/helpers.ts]: renamed it. Good suggestion.

File: test/factories/prismaHook.factory.ts (hunk 3 above)
Comment (mkflow27) [reply test/factories/prismaHook.factory.ts]: Thanks. I have incorporated your suggestion.

File: modules/sor/sorV2/lib/utils/helpers.ts (hunk 1 above)
Comment (mkflow27) [reply modules/sor/sorV2/lib/utils/helpers.ts]: repalced it with `parseEther`. Thanks.

File: yarn.lock
Comment (gmbronco) [inline yarn.lock:1]: we moved to bun for dependency management, so just run bun install instead of yarn install. feel free to delete yarn.lock

**Review by gmbronco — CHANGES_REQUESTED** [review]

File: modules/sor/balancer-sor.integration.test.ts [hunk 5]
```diff
@@ -17,20 +23,24 @@ import { sepolia } from 'viem/chains';
...
- * against the VaultExplorer contract (0x72ebafddc4c7d3eb702c81295d90a8b29f008a03).
+ * against the VaultExplorer contract (0xEB15EBBF9C1a4D7D243d57dE447Df0b97C40c324).
  *
  * TODO: improve test data setup by creating a script that fetches all necessary data automatically for a given blockNumber.
  */
 
+// this test works with 12th deployment.
+// Vault explorer is at 0xEB15EBBF9C1a4D7D243d57dE447Df0b97C40c324
+//
+
```
Comment (brunoguerios) [inline modules/sor/balancer-sor.integration.test.ts]: Can you please remove this and update comments above with deployment 12th?

File: yarn.lock
Comment (mkflow27) [reply yarn.lock:1]: removed the file.

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts [hunk 6]
```diff
@@ -165,9 +195,18 @@ export class StablePool implements BasePoolV3 {
                     kind: RemoveKind.SINGLE_TOKEN_EXACT_IN,
                 },
                 this.poolState,
+                this.hook
```
Comment (brunoguerios) [inline modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts]: Shouldn't `hookState` be passed to swapGivenOut as well?

File: modules/sor/sorV2/lib/poolsV3/weighted/weightedPool.ts [hunk 7]
```diff
@@ -200,6 +240,7 @@ export class WeightedPoolV3 implements BasePoolV3 {
                     swapKind: SwapKind.GivenIn,
                 },
                 this.poolState,
+                this.hook,
```
Comment (brunoguerios) [inline modules/sor/sorV2/lib/poolsV3/weighted/weightedPool.ts]: Same here about adding `hookState` to `swapGivenOut`.
Not sure if it applies, but ideally we should have a test that asserts this is being passed correctly.

File: modules/sor/sorV2/lib/utils/helpers.ts [hunk 8]
```diff
@@ -71,3 +76,46 @@ export function getOutputAmount(paths: PathWithAmount[]): TokenAmount {
     const amounts = paths.map((path) => path.outputAmount);
     return amounts.reduce((a, b) => a.add(b));
 }
+
+export function getHookState(pool: any): HookState | undefined {
```
Comment (brunoguerios) [inline modules/sor/sorV2/lib/utils/helpers.ts:80]: I feel uncomfortable with this `any`, but I think it is what it is 🤷 

@gmbronco - you're the TS magician - any way we can handle this is a safer way? Or is it good enough?

My main issue is with inconsistencies between what we expect and what we get. For example, if hook name is `ExitFee`, we expect it to have a `dynamicData` as attribute. What happens if it's not there for any reason?

Comment (gmbronco) [reply modules/sor/sorV2/lib/utils/helpers.ts:80]: wouldn't PrismaPoolAndHookWithDynamic work? as defined in parent functions.

Comment (mkflow27) [reply modules/sor/sorV2/lib/utils/helpers.ts:80]: Changing it to `PrismaPoolAndHookWithDynamic` introduces build errors

```
modules/sor/sorV2/lib/utils/helpers.ts:87:19 - error TS2339: Property 'name' does not exist on type 'string | number | boolean | JsonObject | JsonArray'.
  Property 'name' does not exist on type 'string'.

87     if (pool.hook.name === 'ExitFee') {
                     ~~~~

modules/sor/sorV2/lib/utils/helpers.ts:89:39 - error TS2339: Property 'dynamicData' does not exist on type 'string | number | boolean | JsonObject | JsonArray'.
  Property 'dynamicData' does not exist on type 'string'.

89         const dynamicData = pool.hook.dynamicData as { removeLiquidityFeePercentage: string };
                                         ~~~~~~~~~~~

modules/sor/sorV2/lib/utils/helpers.ts:98:19 - error TS2339: Property 'name' does not exist on type 'string | number | boolean | JsonObject | JsonArray'.
  Property 'name' does not exist on type 'string'.

98     if (pool.hook.name === 'DirectionalFee') {
                     ~~~~

modules/sor/sorV2/lib/utils/helpers.ts:103:19 - error TS2339: Property 'name' does not exist on type 'string | number | boolean | JsonObject | JsonArray'.
  Property 'name' does not exist on type 'string'.

103     if (pool.hook.name === 'StableSurge') {
                      ~~~~

modules/sor/sorV2/lib/utils/helpers.ts:105:18 - error TS18047: 'pool.typeData' is possibly 'null'.

105             amp: pool.typeData.amp,
                     ~~~~~~~~~~~~~

modules/sor/sorV2/lib/utils/helpers.ts:105:32 - error TS2339: Property 'amp' does not exist on type 'string | number | boolean | JsonObject | JsonArray'.
  Property 'amp' does not exist on type 'string'.

105             amp: pool.typeData.amp,
                                   ~~~

modules/sor/sorV2/lib/utils/helpers.ts:106:50 - error TS18047: 'pool.dynamicData' is possibly 'null'.

106             surgeThresholdPercentage: parseEther(pool.dynamicData.surgeThresholdPercentage),
                                                     ~~~~~~~~~~~~~~~~

modules/sor/sorV2/lib/utils/helpers.ts:106:67 - error TS2339: Property 'surgeThresholdPercentage' does not exist on type 'PrismaPoolDynamicData'.

106             surgeThresholdPercentage: parseEther(pool.dynamicData.surgeThresholdPercentage),
                                                                      ~~~~~~~~~~~~~~~~~~~~~~~~

modules/sor/sorV2/lib/utils/helpers.ts:110:34 - error TS2339: Property 'name' does not exist on type 'string | number | boolean | JsonObject | JsonArray'.
  Property 'name' does not exist on type 'string'.

110     throw new Error(`${pool.hook.name} hook not implemented`);
                                     ~~~~


Found 9 errors in the same file, starting at: modules/sor/sorV2/lib/utils/helpers.ts:87
```

Comment (mkflow27) [reply modules/sor/sorV2/lib/utils/helpers.ts:80]: I'll look more into how I can address the `any`.

Comment (mkflow27) [reply modules/sor/sorV2/lib/utils/helpers.ts:80]: Simply replacing the type with `PrismaPoolAndHookWithDynamic` introduces build errors 

```
modules/sor/sorV2/lib/utils/helpers.ts:87:19 - error TS2339: Property 'name' does not exist on type 'string | number | boolean | JsonObject | JsonArray'.
  Property 'name' does not exist on type 'string'.

87     if (pool.hook.name === 'ExitFee') {
                     ~~~~

modules/sor/sorV2/lib/utils/helpers.ts:89:39 - error TS2339: Property 'dynamicData' does not exist on type 'string | number | boolean | JsonObject | JsonArray'.
  Property 'dynamicData' does not exist on type 'string'.

89         const dynamicData = pool.hook.dynamicData as { removeLiquidityFeePercentage: string };
                                         ~~~~~~~~~~~

modules/sor/sorV2/lib/utils/helpers.ts:98:19 - error TS2339: Property 'name' does not exist on type 'string | number | boolean | JsonObject | JsonArray'.
  Property 'name' does not exist on type 'string'.

98     if (pool.hook.name === 'DirectionalFee') {
                     ~~~~

modules/sor/sorV2/lib/utils/helpers.ts:103:19 - error TS2339: Property 'name' does not exist on type 'string | number | boolean | JsonObject | JsonArray'.
  Property 'name' does not exist on type 'string'.

103     if (pool.hook.name === 'StableSurge') {
                      ~~~~

modules/sor/sorV2/lib/utils/helpers.ts:105:18 - error TS18047: 'pool.typeData' is possibly 'null'.

105             amp: pool.typeData.amp,
                     ~~~~~~~~~~~~~

modules/sor/sorV2/lib/utils/helpers.ts:105:32 - error TS2339: Property 'amp' does not exist on type 'string | number | boolean | JsonObject | JsonArray'.
  Property 'amp' does not exist on type 'string'.

105             amp: pool.typeData.amp,
                                   ~~~

modules/sor/sorV2/lib/utils/helpers.ts:106:50 - error TS18047: 'pool.dynamicData' is possibly 'null'.

106             surgeThresholdPercentage: parseEther(pool.dynamicData.surgeThresholdPercentage),
                                                     ~~~~~~~~~~~~~~~~

modules/sor/sorV2/lib/utils/helpers.ts:106:67 - error TS2339: Property 'surgeThresholdPercentage' does not exist on type 'PrismaPoolDynamicData'.

106             surgeThresholdPercentage: parseEther(pool.dynamicData.surgeThresholdPercentage),
                                                                      ~~~~~~~~~~~~~~~~~~~~~~~~

modules/sor/sorV2/lib/utils/helpers.ts:110:34 - error TS2339: Property 'name' does not exist on type 'string | number | boolean | JsonObject | JsonArray'.
  Property 'name' does not exist on type 'string'.

110     throw new Error(`${pool.hook.name} hook not implemented`);
                                     ~~~~


Found 9 errors in the same file, starting at: modules/sor/sorV2/lib/utils/helpers.ts:87
```

Comment (brunoguerios) [reply modules/sor/sorV2/lib/utils/helpers.ts:80]: Yeah, I don't think it makes sense to use `PrismaPoolAndHookWithDynamic`.
Here we're handling the `hook`  attribute from `PrismaPool`, which is stored as a Json, meaning it can be essentially anything.
I think we should be ok to go with `any` for now and refactor in the future in case we find a better way to handle it @mkflow27

File: modules/sor/balancer-sor.integration.test.ts (hunk 5 above)
Comment (mkflow27) [reply modules/sor/balancer-sor.integration.test.ts]: done.

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts (hunk 6 above)
Comment (mkflow27) [reply modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts]: I have addressed this by renaming variables more clearly.

File: modules/sor/sorV2/lib/poolsV3/weighted/weightedPool.ts (hunk 7 above)
Comment (mkflow27) [reply modules/sor/sorV2/lib/poolsV3/weighted/weightedPool.ts]: I have addressed this by naming the variable more clearly.

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts [hunk 9]
```diff
@@ -58,19 +88,83 @@ describe('SOR V3 Stable Pool Tests', () => {
...
+            tokenRates: tokenRates.map((r) => parseEther(r)),
+            totalSupply: parseEther(totalShares),
+            amp: parseUnits(amp, 3),
+            tokens: tokenAddresses,
+            scalingFactors,
+            aggregateSwapFee
+        };
+        expect(poolState).toEqual(stablePoolWithHook.getPoolState('StableSurge'));
+    })
+    test('results differ when poolState is passed', () => {
```
Comment (brunoguerios) [inline modules/sor/sorV2/lib/poolsV3/stable/stablePool.test.ts]:
```suggestion
    test('results differ when hookState is passed', () => {
```

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts [hunk 10]
```diff
@@ -267,8 +330,16 @@ export class StablePool implements BasePoolV3 {
             scalingFactors: this.tokens.map((t) => t.scalar),
             aggregateSwapFee: 0n,
         };
+
+        if (hookName) {
+            poolState.hookType = hookName;
+        }
```
Comment (brunoguerios) [inline modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts]: You don't need to add the `if` clause - since both are either `string or undefined`, you should be able to assign one to the other directly:
```suggestion
        poolState.hookType = hookName;
```

Aside from that, I saw this is a requirement from balancer-maths, but why have a `hookName` within `poolState`? Intuitively I'd expect it to be inside `hookState` 😅

File: modules/sor/sorV2/lib/poolsV3/weighted/weightedPool.ts [hunk 11]
```diff
@@ -263,8 +323,15 @@ export class WeightedPoolV3 implements BasePoolV3 {
             scalingFactors: this.tokens.map((t) => t.scalar),
             aggregateSwapFee: 0n,
         };
+
+        if (hookName) {
+            poolState.hookType = hookName;
+        }
```
Comment (brunoguerios) [inline modules/sor/sorV2/lib/poolsV3/weighted/weightedPool.ts]: same here

File: modules/sor/balancer-sor.integration.test.ts [hunk 12]
```diff
@@ -122,26 +130,28 @@ describe('Balancer SOR Integration Tests', () => {
...
             });
             const stataDAI = prismaPoolTokenFactory.build({
-                address: '0xde46e43f46ff74a23a65ebb0580cbe3dfe684a17',
-                token: { decimals: 18 },
-                balance: '500',
-                priceRate: '1.101882285912091736',
+                address: '0x80d6d3946ed8a1da4e226aa21ccddc32bd127d1a',
+                token: { decimals: 6 },
+                balance: '1',
+                priceRate: '1.0',
```
Comment (brunoguerios) [inline modules/sor/balancer-sor.integration.test.ts]: Sorry to be picky about this, but I see you updated the token data but didn't change the variable name. I'm assuming 6 decimals is no longer a `stataDAI` 🙃 - do you mind updating the name?

File: modules/sor/balancer-sor.integration.test.ts [hunk 13]
```diff
@@ -423,6 +438,420 @@ describe('Balancer SOR Integration Tests', () => {
         });
     });
 
+    describe('Pools Path with hooks -', async () => {
```
Comment (brunoguerios) [inline modules/sor/balancer-sor.integration.test.ts:436]: Next step is to move some of these tests to a different file. It's starting to get a bit hard to follow this many changes in a single file 😅 
(should be ok to do that in a following PR though)

**Review by brunoguerios — APPROVED** [review]
I left a few minor comments, but otherwise LGTM ✅ 
Let's get this thing out there 🚀

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts [hunk 14]
```diff
@@ -103,6 +103,7 @@ export class StablePool implements BasePoolV3 {
             pool.dynamicData.tokenPairsData as TokenPairData[],
             pool.liquidityManagement,
             hookState,
+            // @ts-ignore: Ignore TypeScript error for accessing name property
             pool.hook?.name,
```
Comment (gmbronco) [inline modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts]: wouldn't it be cleaner to pass the name via the hookState? easier to manage if everything is within one scope.

**Review by gmbronco — CHANGES_REQUESTED** [review]

File: modules/sor/balancer-sor.integration.test.ts (hunk 13 above)
Comment (mkflow27) [reply modules/sor/balancer-sor.integration.test.ts:436]: Created an issue for it #1469

File: modules/sor/balancer-sor.integration.test.ts (hunk 12 above)
Comment (mkflow27) [reply modules/sor/balancer-sor.integration.test.ts]: I have changed the variables

File: modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts (hunk 14 above)
Comment (mkflow27) [reply modules/sor/sorV2/lib/poolsV3/stable/stablePool.ts]: addressed the change request with the latest pushes.

**Review by gmbronco — APPROVED** [review]
